# import GeoJSON data with geopandas

In [1]:
import geopandas as gpd

def checkGeoJson(filename):
    # Path to the GeoJSON file
    geojson_data_path = '/data1/2024_KRISO/yykim/data_ais/' + filename + '.geojson'

    # Read the GeoJSON file
    try:
        gdf_geojson = gpd.read_file(geojson_data_path)
        print("GeoJSON data loaded successfully.")
        display(gdf_geojson.head())  # Display the first few rows of the GeoDataFrame

        return gdf_geojson
    
    except Exception as e:
        print(f"An error occurred while loading the GeoJSON data: {e}")

In [2]:
passenger_gdf = checkGeoJson('passenger_data_edit')
passenger_gdf

GeoJSON data loaded successfully.


SHIP_ID  SEQUENCE_ID           RECPTN_DT         LA          LO   SOG  \
0  123456788            1 2022-08-12 01:39:00  31.468265  126.542485  10.7   
1  123456788            2 2022-08-24 22:00:00  31.409738  126.313397   5.3   
2  123456788            2 2022-08-24 22:01:00  31.410222  126.314067   5.8   
3  123456788            2 2022-08-24 22:02:00  31.411647  126.315073   6.6   
4  123456788            2 2022-08-24 22:03:00  31.413762  126.315803   9.3   

    COG       TYPE   TON        dist  sea_lv port_name  \
0  96.8  passenger  11.0  148.679646   182.0       이어도   
1  60.5  passenger  11.0  133.910851   228.0       이어도   
2  53.0  passenger  11.0  133.930378   228.0       이어도   
3  23.7  passenger  11.0  133.913975   228.0       이어도   
4  13.3  passenger  11.0  133.831449   228.0       이어도   

                     geometry  
0  POINT (125.16667 32.11667)  
1  POINT (125.16667 32.11667)  
2  POINT (125.16667 32.11667)  
3  POINT (125.16667 32.11667)  
4  POINT (125.16667 32.11667)

SHIP_ID  SEQUENCE_ID           RECPTN_DT         LA          LO  \
0         123456788            1 2022-08-12 01:39:00  31.468265  126.542485   
1         123456788            2 2022-08-24 22:00:00  31.409738  126.313397   
2         123456788            2 2022-08-24 22:01:00  31.410222  126.314067   
3         123456788            2 2022-08-24 22:02:00  31.411647  126.315073   
4         123456788            2 2022-08-24 22:03:00  31.413762  126.315803   
...             ...          ...                 ...        ...         ...   
15592619  538004774            2 2022-10-27 20:00:00  33.864178  125.069117   
15592620  538004774            2 2022-10-27 20:01:00  33.860407  125.071595   
15592621  538004774            2 2022-10-27 20:02:00  33.857098  125.074019   
15592622  538004774            2 2022-10-27 20:03:00  33.853790  125.076443   
15592623  538004774            2 2022-10-27 20:04:00  33.850482  125.078867   

                SOG    COG       TYPE      TON        dist  sea_lv port_name  \
0         10.700000   96.8  passenger     11.0  148.679646   182.0       이어도   
1          5.300000   60.5  passenger     11.0  133.910851   228.0       이어도   
2          5.800000   53.0  passenger     11.0  133.930378   228.0       이어도   
3          6.600000   23.7  passenger     11.0  133.913975   228.0       이어도   
4          9.300000   13.3  passenger     11.0  133.831449   228.0       이어도   
...             ...    ...        ...      ...         ...     ...       ...   
15592619  13.200000  151.2  passenger  43673.0   21.122813    48.0       가거도   
15592620  13.100000  150.3  passenger  43673.0   21.487866    48.0       가거도   
15592621  13.133333  149.6  passenger  43673.0   21.807479    48.0       가거도   
15592622  13.166667  148.9  passenger  43673.0   22.130848    48.0       가거도   
15592623  13.200000  148.2  passenger  43673.0   22.457809    48.0       가거도   

                            geometry  
0         POINT (125.16667 32.11667)  
1         POINT (125.16667 32.11667)  
2         POINT (125.16667 32.11667)  
3         POINT (125.16667 32.11667)  
4         POINT (125.16667 32.11667)  
...                              ...  
15592619  POINT (125.11667 34.05000)  
15592620  POINT (125.11667 34.05000)  
15592621  POINT (125.11667 34.05000)  
15592622  POINT (125.11667 34.05000)  
15592623  POINT (125.11667 34.05000)  

[15592624 rows x 13 columns]

In [4]:
cargo_gdf = checkGeoJson('cargo_data_edit')
cargo_gdf

GeoJSON data loaded successfully.


SHIP_ID  SEQUENCE_ID           RECPTN_DT         LA          LO   SOG  \
0     123            1 2022-07-25 05:34:00  34.931833  127.774000  5.10   
1     123            1 2022-07-25 05:35:00  34.929833  127.776167  5.60   
2     123            1 2022-07-25 05:36:00  34.929667  127.776333  5.60   
3     123            1 2022-07-25 05:37:00  34.928000  127.778417  5.95   
4     123            1 2022-07-25 05:38:00  34.926333  127.780500  6.30   

     COG   TYPE     TON      dist  sea_lv port_name  \
0  137.0  cargo  1241.0  4.161515   247.0        광양   
1  134.0  cargo  1241.0  4.086203   248.0        광양   
2  133.0  cargo  1241.0  4.080098   248.0        광양   
3  133.5  cargo  1241.0  4.050591   248.0        광양   
4  134.0  cargo  1241.0  4.038350   248.0        광양   

                     geometry  
0  POINT (127.75000 34.90000)  
1  POINT (127.75000 34.90000)  
2  POINT (127.75000 34.90000)  
3  POINT (127.75000 34.90000)  
4  POINT (127.75000 34.90000)

SHIP_ID  SEQUENCE_ID           RECPTN_DT         LA          LO  \
0               123            1 2022-07-25 05:34:00  34.931833  127.774000   
1               123            1 2022-07-25 05:35:00  34.929833  127.776167   
2               123            1 2022-07-25 05:36:00  34.929667  127.776333   
3               123            1 2022-07-25 05:37:00  34.928000  127.778417   
4               123            1 2022-07-25 05:38:00  34.926333  127.780500   
...             ...          ...                 ...        ...         ...   
98344030  750226000            2 2022-10-27 15:50:00  33.757393  126.283013   
98344031  750226000            3 2022-10-27 20:00:00  33.800577  125.660253   
98344032  750226000            3 2022-10-27 20:01:00  33.800832  125.657871   
98344033  750226000            3 2022-10-27 20:02:00  33.801088  125.655489   
98344034  750226000            3 2022-10-27 20:03:00  33.801343  125.653107   

               SOG         COG   TYPE     TON       dist  sea_lv port_name  \
0         5.100000  137.000000  cargo  1241.0   4.161515   247.0        광양   
1         5.600000  134.000000  cargo  1241.0   4.086203   248.0        광양   
2         5.600000  133.000000  cargo  1241.0   4.080098   248.0        광양   
3         5.950000  133.500000  cargo  1241.0   4.050591   248.0        광양   
4         6.300000  134.000000  cargo  1241.0   4.038350   248.0        광양   
...            ...         ...    ...     ...        ...     ...       ...   
98344030  7.900000  263.700000  cargo   499.0  21.474280   181.0       추자도   
98344031  6.900000  276.100000  cargo   499.0  47.864919    40.0       만재도   
98344032  6.633333  276.433333  cargo   499.0  47.757059    41.0       만재도   
98344033  6.366667  276.766667  cargo   499.0  47.649984    41.0       만재도   
98344034  6.100000  277.100000  cargo   499.0  47.543699    41.0       만재도   

                            geometry  
0         POINT (127.75000 34.90000)  
1         POINT (127.75000 34.90000)  
2         POINT (127.75000 34.90000)  
3         POINT (127.75000 34.90000)  
4         POINT (127.75000 34.90000)  
...                              ...  
98344030  POINT (126.30000 33.95000)  
98344031  POINT (125.46667 34.20000)  
98344032  POINT (125.46667 34.20000)  
98344033  POINT (125.46667 34.20000)  
98344034  POINT (125.46667 34.20000)  

[98344035 rows x 13 columns]

In [8]:
passenger_gdf.describe()

SEQUENCE_ID                      RECPTN_DT            LA  \
count  1.559262e+07                       15592624  1.559262e+07   
mean   1.203989e+02  2022-08-30 03:23:30.021364992  3.540867e+01   
min    1.000000e+00            2022-06-01 00:00:00  3.100007e+01   
25%    4.100000e+01            2022-07-15 10:36:00  3.447328e+01   
50%    9.300000e+01            2022-08-28 10:20:00  3.483649e+01   
75%    1.670000e+02            2022-10-15 14:12:00  3.673278e+01   
max    9.390000e+02            2022-11-30 23:59:00  3.899993e+01   
std    1.067315e+02                            NaN  1.279984e+00   

                 LO           SOG           COG           TON          dist  \
count  1.559262e+07  1.559262e+07  1.559262e+07  1.559262e+07  1.559262e+07   
mean   1.272009e+02  1.278545e+01  1.773781e+02  4.482845e+03  1.496998e+01   
min    1.240000e+02  5.100000e+00  0.000000e+00  5.430000e+00  6.893607e-04   
25%    1.262341e+02  7.800000e+00  8.954430e+01  1.450000e+02  4.464883e+00   
50%    1.265966e+02  1.015385e+01  1.829333e+02  3.760000e+02  8.180479e+00   
75%    1.283893e+02  1.280000e+01  2.573000e+02  2.374000e+03  1.403115e+01   
max    1.320000e+02  1.023000e+02  4.095000e+02  2.510530e+05  4.666608e+02   
std    1.444803e+00  1.223186e+01  9.957996e+01  1.206810e+04  2.274561e+01   

             sea_lv  
count  1.553805e+07  
mean   2.208588e+02  
min   -5.200000e+01  
25%    9.400000e+01  
50%    1.840000e+02  
75%    2.940000e+02  
max    9.630000e+02  
std    1.737837e+02

In [9]:
cargo_gdf.describe()

SEQUENCE_ID                      RECPTN_DT            LA  \
count  9.834404e+07                       98344035  9.834404e+07   
mean   3.631702e+01  2022-08-30 06:33:19.699393280  3.475051e+01   
min    1.000000e+00            2022-06-01 00:00:00  3.100000e+01   
25%    5.000000e+00            2022-07-16 02:49:00  3.400035e+01   
50%    1.600000e+01            2022-08-29 22:06:00  3.464184e+01   
75%    4.600000e+01            2022-10-14 00:27:00  3.548120e+01   
max    7.630000e+02            2022-11-30 23:59:00  3.900000e+01   
std    5.330232e+01                            NaN  1.398750e+00   

                 LO           SOG           COG           TON          dist  \
count  9.834404e+07  9.834404e+07  9.834404e+07  9.834404e+07  9.834404e+07   
mean   1.275585e+02  1.189274e+01  1.671220e+02  2.470292e+04  4.554784e+01   
min    1.240000e+02  5.100000e+00  0.000000e+00  2.900000e+01  1.810513e-03   
25%    1.259130e+02  9.500000e+00  6.580000e+01  2.944000e+03  1.357237e+01   
50%    1.276789e+02  1.120000e+01  1.810000e+02  7.726000e+03  2.899712e+01   
75%    1.291800e+02  1.300000e+01  2.459500e+02  3.241500e+04  6.291734e+01   
max    1.320000e+02  1.023000e+02  4.095000e+02  1.222166e+06  4.947581e+02   
std    1.884409e+00  7.695231e+00  1.040125e+02  3.680459e+04  4.502050e+01   

             sea_lv  
count  9.830893e+07  
mean   1.589904e+02  
min   -5.000000e+01  
25%    6.000000e+01  
50%    1.340000e+02  
75%    2.150000e+02  
max    9.630000e+02  
std    1.317011e+02

- passenger: 15592624 rows × 13 columns
- cargo: 98344035 rows × 13 columns

- SOG (Speed over ground): 0.1 knot resolution from 0 to 102 knots
  - AIS specification for sog (Speed over ground) shows that 102.3 knots is reported when the vessel speed is unavailable
 https://documentation.spire.com/ais-fundamentals/speed-over-ground-sog-rules/
  - **1 knot = 1nm / 1hour**  (nm: nautical mile, 1nm = 1' latitude = 1.852 km)
- COG (Course over ground): relative to true north to 0.1 degree
---

# Collision Prevention: Calculation
1. input: SHIP_ID, RECPTN_DT(datetime), time length (optional)
2. calculate circular area of focus: according to SHIP_ID, RECPTN_DT
  - center: Latitude and Longitude point of the SHIP_ID at RECPTN_DT
  - radius: SOG value of the vessel of the SHIP_ID at RECPTN_DT (default)
    - if other specific time length is mentioned in input, calculate new radius
    - e.g. 40 min: radius = SOG value * (40/60)
3. split the circular area into 6 parts: according to COG
  - coordinates axis: COG
  - area1: +30 degrees & -30 degrees from the coordinate axis (60 degree arc)
  - area2 ~ 6: 60 degree arc each, in clockwise direction of area1
4. calculate the number of all the other ships in each area: according to RECPTN_DT
  - timeline to focus: 1 hour from input RECPTN_DT (default)
    - if other specific time length is mentioned in input, calculate new timeline
    - e.g. 40 min: 2022-08-24 19:09:00 ~ 2022-08-24 19:49:00
  - add up the numbers of unique ships that pass through each area at least once
    - e.g. ship with id '308628000' is detected in area 2 during the timestamp: add 1 count to area2
    - if a same ship passes multiple areas during the expectied timeline, add 1 count for each area
    - e.g. ship with id '308628000' is detected in area 2,3,4 during the timestamp: add 1 count to area2,3,4 each
5. return the unique ship counts and ship ids for each area
  - e.g. {area1: {0,[]}, area2: {2, ['229659000','308628000']}, area3: {2, ['229659000','308628000']}, area4: {1, ['308628000']}, area5: {1, ['209682000']}, area6: {3, ['123456788','209682000','229416000']}}

In [3]:
import geopandas as gpd
from shapely.geometry import Point, Polygon
import pandas as pd
from datetime import timedelta
import math
import numpy as np

'''
# Function to read the GeoJSON file
def checkGeoJson(filename):
    # Path to the GeoJSON file
    geojson_data_path = '/data1/2024_KRISO/yykim/data_ais/' + filename + '.geojson'
    
    # Read the GeoJSON file
    try:
        gdf_geojson = gpd.read_file(geojson_data_path)
        print("GeoJSON data loaded successfully.")
        display(gdf_geojson.head())  # Display the first few rows of the GeoDataFrame
        return gdf_geojson
    except Exception as e:
        print(f"An error occurred while loading the GeoJSON data: {e}")
'''

# Function to calculate ships around an input ship
def calculate_ships_around(df, ship_id, recptn_dt, time_length=60):
    # Ensure RECPTN_DT is datetime
    df['RECPTN_DT'] = pd.to_datetime(df['RECPTN_DT'])
    
    # Get the specific row of the input ship at the given timestamp
    ship_row = df[(df['SHIP_ID'] == ship_id) & (df['RECPTN_DT'] == recptn_dt)].iloc[0]
    ship_position = Point(ship_row['LO'], ship_row['LA'])
    sog = ship_row['SOG']
    cog = ship_row['COG']
    
    # Calculate radius based on SOG and time length
    radius = sog * (time_length / 60)
    
    # Define the time window
    start_time = recptn_dt - timedelta(minutes=time_length / 2)
    end_time = recptn_dt + timedelta(minutes=time_length / 2)
    
    # Filter ships within the time window
    filtered_df = df[(df['RECPTN_DT'] >= start_time) & (df['RECPTN_DT'] <= end_time)]
    
    # Function to create circular sector
    def create_sector(center, radius, start_angle, end_angle, num_points=100):
        angles = np.linspace(start_angle, end_angle, num_points)
        points = [
            Point(
                center.x + radius * math.cos(math.radians(angle)),
                center.y + radius * math.sin(math.radians(angle))
            )
            for angle in angles
        ]
        return Polygon([center] + points + [center])
    
    # Create 6 sectors
    sectors = []
    for i in range(6):
        start_angle = cog - 30 + i * 60
        end_angle = start_angle + 60
        sectors.append(create_sector(ship_position, radius, start_angle, end_angle))
    
    # Count ships in each sector
    result = {}
    for i, sector in enumerate(sectors):
        area_ships = filtered_df[filtered_df['geometry'].apply(lambda x: sector.contains(x))]
        unique_ships = area_ships['SHIP_ID'].unique()
        result[f'area{i+1}'] = {
            'count': len(unique_ships),
            'ship_ids': unique_ships.tolist()
        }
    
    return result

## function test

In [4]:
passenger_gdf.SHIP_ID.unique()[0:10]

array(['123456788', '209682000', '210326000', '229166000', '229416000',
       '229659000', '232036828', '256334000', '273199300', '308628000'],
      dtype=object)

In [6]:
passenger_gdf[passenger_gdf.SHIP_ID == '232036828']

SHIP_ID  SEQUENCE_ID           RECPTN_DT         LA          LO  \
13784  232036828            1 2022-08-24 19:07:00  36.074563  124.039223   
13785  232036828            1 2022-08-24 19:08:00  36.070978  124.040422   
13786  232036828            1 2022-08-24 19:09:00  36.067523  124.041751   
13787  232036828            1 2022-08-24 19:10:00  36.064068  124.043080   
13788  232036828            1 2022-08-24 19:11:00  36.060613  124.044409   
...          ...          ...                 ...        ...         ...   
20279  232036828           10 2022-10-31 10:46:00  35.583331  130.582471   
20280  232036828           10 2022-10-31 10:47:00  35.585395  130.585876   
20281  232036828           10 2022-10-31 10:48:00  35.587458  130.589280   
20282  232036828           10 2022-10-31 10:49:00  35.589522  130.592685   
20283  232036828           10 2022-10-31 10:50:00  35.591585  130.596090   

             SOG         COG       TYPE      TON        dist  sea_lv  \
13784  13.500000  165.000000  passenger  43672.0  161.541446   165.0   
13785  13.400000  164.200000  passenger  43672.0  161.869579   164.0   
13786  13.400000  163.891304  passenger  43672.0  162.180271   164.0   
13787  13.400000  163.582609  passenger  43672.0  162.491359   163.0   
13788  13.400000  163.273913  passenger  43672.0  162.802842   162.0   
...          ...         ...        ...      ...         ...     ...   
20279  12.517391   53.026087  passenger  43672.0  101.237817    19.0   
20280  12.513043   52.969565  passenger  43672.0  101.457799    19.0   
20281  12.508696   52.913043  passenger  43672.0  101.678738    19.0   
20282  12.504348   52.856522  passenger  43672.0  101.900626    19.0   
20283  12.500000   52.800000  passenger  43672.0  102.123456    19.0   

      port_name                    geometry  
13784     옹진소청초  POINT (124.73333 37.41667)  
13785     옹진소청초  POINT (124.73333 37.41667)  
13786     옹진소청초  POINT (124.73333 37.41667)  
13787     옹진소청초  POINT (124.73333 37.41667)  
13788     옹진소청초  POINT (124.73333 37.41667)  
...         ...                         ...  
20279       양포항  POINT (129.51667 35.86667)  
20280       양포항  POINT (129.51667 35.86667)  
20281       양포항  POINT (129.51667 35.86667)  
20282       양포항  POINT (129.51667 35.86667)  
20283       양포항  POINT (129.51667 35.86667)  

[6500 rows x 13 columns]

In [4]:
# Calculate ships around the input ship
ship_id = '232036828'  # Example ship ID
recptn_dt = pd.to_datetime('2022-08-24 19:11:00')  # Example timestamp
result = calculate_ships_around(passenger_gdf, ship_id, recptn_dt)
print(result)

{'area1': {'count': 0, 'ship_ids': []}, 'area2': {'count': 0, 'ship_ids': []}, 'area3': {'count': 1, 'ship_ids': ['667002108']}, 'area4': {'count': 54, 'ship_ids': ['232036828', '351280000', '352001378', '352001576', '354085000', '431602232', '440001130', '440002010', '440009210', '440015380', '440016380', '440018420', '440026000', '440029550', '440041740', '440059890', '440075110', '440077110', '440104800', '440105630', '440108320', '440109400', '440113620', '440116240', '440118560', '440121350', '440121700', '440123380', '440124440', '440126940', '440132930', '440139880', '440182610', '440190220', '440193260', '440207640', '440218810', '440223310', '440302330', '440304410', '440307870', '440319130', '440320190', '440329260', '440333660', '440334250', '440335170', '440404100', '440413000', '440414910', '440600447', '441198000', '441368000', '441877000']}, 'area5': {'count': 13, 'ship_ids': ['232036828', '351735000', '355925000', '374869000', '440006070', '440009170', '440061040', '440

---

In [4]:
import pandas as pd

# Path to the JSON file
json_data_path = '/data1/2024_KRISO/yykim/data_ais/passenger_data_edit.json'

# Read the JSON file
try:
    df_json = pd.read_json(json_data_path)
    print("JSON data loaded successfully.")
    display(df_json.head())  # Display the first few rows of the dataframe
except Exception as e:
    print(f"An error occurred while loading the JSON data: {e}")

JSON data loaded successfully.


SHIP_ID       TYPE       TON  \
0  123456788  passenger      11.0   
1  209682000  passenger   44095.0   
2  210326000  passenger   48035.0   
3  229166000  passenger   91994.0   
4  229416000  passenger  130200.0   

                                           SEQUENCES  
0  [{'SEQUENCE_ID': 1, 'SEQ_PROPERTIES': [{'SHIP_...  
1  [{'SEQUENCE_ID': 1, 'SEQ_PROPERTIES': [{'SHIP_...  
2  [{'SEQUENCE_ID': 1, 'SEQ_PROPERTIES': [{'SHIP_...  
3  [{'SEQUENCE_ID': 1, 'SEQ_PROPERTIES': [{'SHIP_...  
4  [{'SEQUENCE_ID': 1, 'SEQ_PROPERTIES': [{'SHIP_...